### ISSUE TITLE
Container remains in rotation for service virtual IP during update

### ISSUE THREAD

**Issue by evanp** [docker/docker#26672]
**Description**

While updating a service, the currently updating task remains in the rotation for the virtual IP during the update. This causes failed requests for 1/N of the clients of that service.

**Steps to reproduce the issue:**

1. Start a service with N replicas and `update-parallelism M` where M < N, so that some tasks stay alive during the update.
2. Update the service with a new image (`docker service update --image <image name> <service name>`)
3. While the service is updating, make an HTTP or other request to the service virtual IP.

**Describe the results you received:**

ECONNREFUSED on the virtual IP for an HTTP request from a client within the swarm, as follows:

'http://home/dist/bundle.js is down: connect ECONNREFUSED 10.0.23.102:80.'

This happens about 1/N times, which would be expected if the task is still in the rotation for the virtual IP.

**Describe the results you expected:**

I expected one of the other tasks, still running, to be called instead.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 8
 Running: 4
 Paused: 0
 Stopped: 4
Images: 25
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 191
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: active
 NodeID: cywcp04msr1dazdc8o8yv27pd
 Is Manager: true
 ClusterID: dpffe5ss6yhpgixuuhwjx2qtw
 Managers: 3
 Nodes: 31
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.134.37.79
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-36-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: prod1
ID: WPI3:OJIY:UPNC:7I66:BOOT:3XHD:RBPG:AIND:4BUM:BXZ4:7UOR:46SE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: fuzzyioproduction
Registry: https://index.docker.io/v1/
Labels:
WARNING: No swap limit support
 provider=digitalocean
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running on digital ocean.

**Comment by thaJeztah** [comment]
Thanks for reporting. This may be a duplicate (and possibly fixed already), but not sure which issue :smile:

ping @mrjana perhaps you know

**Comment by evanp** [comment]
OK. I just grepped the virtual IP address in the syslog for the master node; here are some interesting results:

```
Sep 17 19:45:34 prod1 docker[12089]: time="2016-09-17T19:45:34.290492420Z" level=error msg="Failed to create real server 10.0.23.79 for vip 10.0.23.102 fwmark 356 in sb 5118178e7578f03d3612ea824410ea11a636472a80cb4a79595ddeb8f0f8e172: no such process"
Sep 17 19:45:34 prod1 docker[12089]: time="2016-09-17T19:45:34.297227889Z" level=error msg="Failed to create real server 10.0.23.73 for vip 10.0.23.102 fwmark 356 in sb 5118178e7578f03d3612ea824410ea11a636472a80cb4a79595ddeb8f0f8e172: no such process"
Sep 17 19:46:05 prod1 docker[12089]: time="2016-09-17T19:46:05.988541775Z" level=error msg="Failed to create real server 10.0.23.78 for vip 10.0.23.102 fwmark 369 in sb b1423c2236f6c8567dc978278c199522908f04c5f6b08a6efdadb8b0612d0039: no such process"
Sep 17 19:46:06 prod1 docker[12089]: time="2016-09-17T19:46:06.651944733Z" level=error msg="Failed to create real server 10.0.23.74 for vip 10.0.23.102 fwmark 369 in sb b1423c2236f6c8567dc978278c199522908f04c5f6b08a6efdadb8b0612d0039: no such process"
Sep 17 19:46:16 prod1 docker[12089]: time="2016-09-17T19:46:16.105362643Z" level=error msg="Failed to create real server 10.0.23.78 for vip 10.0.23.102 fwmark 373 in sb d7917d4a3144ea670189eef0127523261866bd3602d91c3f63c6516ff9794910: no such process"
Sep 17 19:46:16 prod1 docker[12089]: time="2016-09-17T19:46:16.706891346Z" level=error msg="Failed to create real server 10.0.23.74 for vip 10.0.23.102 fwmark 373 in sb d7917d4a3144ea670189eef0127523261866bd3602d91c3f63c6516ff9794910: no such process"
Sep 17 19:46:16 prod1 docker[12089]: time="2016-09-17T19:46:16.796861969Z" level=info msg="Failed to delete real server 10.0.23.73 for vip 10.0.23.102 fwmark 373: no such process"
Sep 17 19:46:36 prod1 docker[12089]: time="2016-09-17T19:46:36.869235302Z" level=error msg="Failed to create real server 10.0.23.74 for vip 10.0.23.102 fwmark 382 in sb a66bab36417074119d69340228f9b95b5b890e4919ed0750acd46985f56d2328: no such process"
Sep 17 19:46:37 prod1 docker[12089]: time="2016-09-17T19:46:37.195660718Z" level=error msg="Failed to create real server 10.0.23.78 for vip 10.0.23.102 fwmark 382 in sb a66bab36417074119d69340228f9b95b5b890e4919ed0750acd46985f56d2328: no such process"
```

**Comment by xiaods** [comment]
@evanp cloud you please test it based on upstream docker binary.

**Comment by evanp** [comment]
@xiads It's not clear what you mean. I'm using Docker 1.12.1 installed via Docker Machine.

**Comment by xiaods** [comment]
@evanp i found the @thaJeztah mentioned this report maybe a duplicate bug fixed already. so i suggest you can build a docker binary from upstream repo, then reproduce your steps again.

**Comment by mrjana** [comment]
@evanp I don't know if you are running into any specific issue since our load balancing state is eventually consistent there is always a time window while your service is getting updated that there are some stale backends. But this time window should be very very minimal. So unless you seeing the old task in the backend for a prolonged time it is expected. If it is indeed for a prolonged period of time then I would suggest trying out the docker master code to make sure if that issue is fixed.

Having said that, one improvement that we are probably going to make in 1.13 is that we are hooking up the load balancing state to the health of the task. In a case of a rolling update, we would first remove the task from load balancer backend state, and then stop and remove the task and the create a new task for the updated service. Add the new task to the load balancer state only after making sure that the task started healthy. This will ensure that when a rolling update happens your service undergoes minimal disruption.

**Comment by aluzzardi** [comment]
This is a duplicate of #24092 which will be addressed in 1.13.

@evanp correct? Closing this for now, reopen if I'm mistaken

**Comment by evanp** [comment]
@mrjana so, good news bad news.

Bad news is that we're still seeing this issue in 1.13-dev. However, they're survivable. The work on #24092 will make these upgrades much more seamless.
